core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -717,6 +717,19 @@ public TableMetadata removeSnapshotLogEntries(Set<Long> snapshotIds) {
         snapshots, newSnapshotLog, addPreviousFile(file, lastUpdatedMillis));
   }
 
+  public TableMetadata withUpdatedSnapshotAndSchema(long snapshotId, int schemaId, int sortOrderId, int specId) {
```

Comment:
I think that things would be easier with the changes from https://github.com/apache/iceberg/pull/2957, where we could just call the builder and only update the actual fields that we need. Then we wouldn't need to have the `withUpdatedSnapshotAndSchema` method in `TableMetadata` and could do everything in the `NessieCatalog` where we need it. Maybe we should go forward with this as is right now and then refactor/improve this once #2957 is merged?

Reply:
For #2957 I was waiting for more clarity around the REST catalog to design the table metadata update builder semantics, so it probably won't be merged very soon. But I agree with Ryan that this seems like a very restrictive method, and the intention of the operation is unclear comparing to the other update methods (although I know how it fits in the Nessie context).

To update the schemaId, sortOrderId, specId to a specific version, I think it makes more sense to expose specific APIs in related update classes instead of having everything done through a separated method. What do you think?